ibis/expr/types/joins.py

```diff
@@ -0,0 +1,191 @@
+        else:
+            if isinstance(pred, tuple):
+                if len(pred) != 2:
+                    raise ExpressionError("Join key tuple must be length 2")
+                lk, rk = pred
+            else:
+                lk = rk = pred
+
+            # bind the predicates to the join chain
+            (left_value,) = bind(left, lk)
+            (right_value,) = bind(right, rk)
+
+            # dereference the left value to one of the relations in the join chain
+            left_value = left_value.op().replace(deref_left, filter=ops.Value)
+            right_value = right_value.op().replace(deref_right, filter=ops.Value)
+
+            yield ops.Equals(left_value, right_value).to_expr()
+
+
+@public
+class JoinExpr(TableExpr):
+    def join(
+        self,
+        right,
+        predicates: Any,
+        # TODO(kszucs): add typehint about the possible join kinds
+        how: str = "inner",
+        *,
+        lname: str = "",
+        rname: str = "{name}_right",
+    ):
+        """Join with another table."""
+        import pyarrow as pa
+        import pandas as pd
+
+        if isinstance(right, (pd.DataFrame, pa.Table)):
+            right = ibis.memtable(right)
+        elif not isinstance(right, TableExpr):
+            raise TypeError(
+                f"right operand must be a TableExpr, got {type(right).__name__}"
+            )
+
+        left = self.op()
+        right = right.op()
+        if isinstance(right, ops.SelfReference):
+            deref_right = {}
+        else:
+            right = ops.SelfReference(right)
+            deref_right = {v: ops.Field(right, k) for k, v in right.values.items()}
+
+        # construct the various dereference mappings
+        deref_left = {ops.Field(left, k): v for k, v in left.values.items()}
+        deref_both = deref_left.copy()
+        for table in self._relations():
+            deref_both.update({v: ops.Field(table, k) for k, v in table.values.items()})
+        # finally add the right dereferences to take precedence
+        deref_both.update(deref_right)
+
+        # bind and dereference the predicates
+        preds = prepare_predicates(
+            left.to_expr(),
+            right.to_expr(),
+            predicates,
+            deref_left=deref_left,
+            deref_right=deref_right,
+            deref_both=deref_both,
+        )
+        preds = flatten_predicates(list(preds))
+
+        # calculate the fields based in lname and rname, this should be a best
+        values = disambiguate_fields(how, left.values, right.fields, lname, rname)
+
+        # construct a new join link and add it to the join chain
+        link = ops.JoinLink(how, table=right, predicates=preds)
+        left = left.copy(rest=left.rest + (link,), values=values)
+
+        # return with a new JoinExpr wrapping the new join chain
+        return self.__class__(left)
+
+    def select(self, *args, **kwargs):
+        """Select expressions."""
```

Comment:
Before we merge, let's copy the `Table.select` signature and docstring here and do the same for the rest of the APIs.

Reply:
I could use some suggestions here. Actually we should override `__getattribute__` to ensure that `self.finish()` gets called in all methods. Any suggestions for that?

Reply:
Done.